glmmTMB/R/glmmTMB.R

```diff
@@ -78,12 +80,36 @@ mkTMBStruc <- function(formula, ziformula, dispformula,
 
   if (is.null(weights)) weights <- rep(1, nobs)
 
-  ## binomial family: At this point we know that (yobs, weights) are
-  ## (proportions, size) as output from binomial()$initialize.
-  ## On the C++ side 'yobs' must be the actual observations (counts).
+  ## binomial family:
+  ## binomial()$initialize was only executed locally  
+  ## yobs could be a factor -> treat as binary following glm
+  ## yobs could be cbind(success, failure)
+  ## yobs could be binary
+  ## (yobs, weights) could be (proportions, size)
+  ## On the C++ side 'yobs' must be the number of successes.
   if ( binomialType(family$family) ) {
-    yobs <- weights * yobs
+    if (is.factor(yobs)) {
+      ## following glm, ‘success’ is interpreted as the factor not
+      ## having the first level (and hence usually of having the
+      ## second level).
+      yobs <- pmin(as.numeric(yobs)-1,1)
+      size <- rep(1, nobs)
+    } else {
+      if(is.matrix(yobs)) { # yobs=cbind(success, failure)
+        size <- yobs[,1] + yobs[,2]
+        yobs <- yobs[,1] #successes
+      } else {
+      if(all(yobs %in% c(0,1))) { #binary
+        size <- rep(1, nobs)
+      } else { #proportions
+          yobs <- weights * yobs
+          size <- weights
+          weights <- rep(1, nobs)
+        }
+      }
+    }
   }
+  if (is.null(size)) size <- rep(1, nobs)
```

Comment:
Better to pass `numeric(0)` as default `size`. That'll ensure it's not used somewhere by accident.